src/Access/RoleCache.cpp

```diff
@@ -163,11 +163,15 @@ void RoleCache::roleChanged(const UUID & role_id, const RolePtr & changed_role)
 
     std::lock_guard lock{mutex};
     auto role_from_cache = cache.get(role_id);
-    if (!role_from_cache)
-        return;
-    role_from_cache->first = changed_role;
-    cache.update(role_id, role_from_cache);
-    collectEnabledRoles(&notifications);
```

Comment:
Actually it seems this fix is not enough. Because if a role expires from the cache not only it got deleted from the cache but the subscription got deleted too. And that's bad because some `EnabledRoles` might still depends on that role. Apparently we need to store subscriptions in `EnabledRoles` and not in this role cache.

Reply:
@ilejn Do you want to finish this PR?

Reply:
Also it would be nice to write a test and make the timeout of that cache configurable.

Reply:
Hello @vitibar
1. From my point of view adding a role back to cache is a valid plan. Am I wrong? The behaviour seems logical: a role is removed from cache while nothing happens; upon the role change we put it back to the cache. 
1a. On the contrary, deleting subscriptions on a recently changed role is not perfect.
2. Yes, I willingly go on working on this thing. Though I might need to discuss some design approaches with you.
3. I'll create the test. 
3a. Not sure that introducing a configurations parameters purely to facilitate testing is a perfect idea, though we can discuss it as well.

Reply:
1. We don't need it back. I know the class is named `RoleCache` but it's not actually a cache of roles themselves, it's more like a cache of enabled roles. If a role is in this cache it means it's used by some users. If a role expires and gets deleted from the cache that means it's not in use anymore. So it doesn't matter in this case whether the role has been changed recently or not.
3. If the timeout becomes configurable it will help to write the test (no normal test should wait 10 minutes, that's too long).
```

Reply:
About the extra required change which I told before - we need to keep our subscriptions from destruction for all enabled roles. We can do that by changing the types of two variables:

```
    Poco::AccessExpireCache<UUID, std::pair<RolePtr, std::shared_ptr<scope_guard>>> cache;

    struct EnabledRolesWithSubscriptions
    {
        std::weak_ptr<EnabledRoles> enabled_roles;
        std::vector<std::shared_ptr<std::shared_ptr<scope_guard>>> subscriptions;
    };
    std::map<EnabledRoles::Params, EnabledRolesWithSubscriptions> enabled_roles;
```

And after `getRole()` gets a role a subscription should be put not only to `cache` but also to `EnabledRolesWithSubscriptions::subscriptions`. If we do that we will be sure we'll never lose any subscriptions to any role in use.

Reply:
Ok, it looks like I understand my mistake: I've played with a single role.
While we have only one role it is safe to recreate subscription from RoleCache::roleChanged. 
If we have more roles it is possible that role A can be purged form cache in the context of process role B, and A's subscriptions are lost.
Does it make sense?